### ISSUE TITLE
bigip_command got REST "maximum active login tokens" 400 error

### ISSUE THREAD

**Issue by earchibald** [F5Networks/f5-ansible#269]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
bigip_command

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### PYTHON VERSION
```
Python 2.7.5
```

##### BIGIP VERSION
```
Sys::Version
Main Package
  Product     BIG-IP
  Version     13.0.0
  Build       2.0.1671
  Edition     Hotfix HF2
  Date        Fri Apr 28 20:36:43 PDT 2017

Hotfix List
ID643325      ID561592    ID642015-4   ID552988-1  ID644805    ID643547-2
ID636520-1    ID639530    ID636699-6   ID640376-2  ID641612-1  ID596924-1
ID569100      ID617901-9  ID641445-2   ID598002-9  ID603609-1  ID600205-1
ID575642      ID650450    ID648990-1   ID633091    ID643332    ID598024-1
ID641547-1    ID622160    ID642221-1   ID638780    ID620759-3  ID647962-1
ID643396-1    ID590091-4  ID638629-1   ID429213    ID642124    ID630712
ID629921-3    ID636675    ID636790-4   ID609995-1  ID570855-1  ID605792
ID642330      ID541550-2  ID643330     ID642119-1  ID629017    ID643327
ID638799-2    ID644817-1  ID559080     ID629752    ID638115-1  ID594228-1
ID641390-2    ID649933-2  ID618430-1   ID247527-1  ID636702-4  ID569814
ID633564-1    ID636289-1  ID644870     ID632499-2  ID643187-1  ID619593-1
ID633879-2    ID366695-9  ID561596     ID635111-1  ID640824-2  ID645805-2
ID517756-5    ID582773    ID627747     ID615267    ID611161-4  ID639500-1
ID642058      ID636663    ID606710-11  ID660170-2  ID638825-1  ID639630-1
ID654368-1    ID608245-1  ID417720     ID635754-2  ID610307-4  ID570926
ID635189-1    ID641724    ID636479     ID645203-1  ID620625-3  ID615372
ID636853-4    ID638967-2  ID578017     ID638215    ID651627    ID639764-1
ID639528      ID642874-2  ID588414-1   ID642400-3  ID631572    ID620659-4
ID649048      ID626386-2  ID649369-1   ID621974-5  ID610485-1  ID637847
ID644855-1    ID629013    ID639729-1   ID637141-1  ID645219    ID643326
ID658764      ID648867-1  ID642211-1   ID639383    ID639565    ID639049-1
ID633985-1    ID578076    ID637181-1   ID623362-1  ID639767-1  ID639774-1
ID541320-8    ID639744-3  ID635688     ID652151    ID649617-1  ID636155
ID621379-1    ID641482-3  ID630610-1   ID563165    ID643404-1  ID627695-1
ID644490-2    ID624722    ID639288-1   ID646719-1  ID648056-3  ID640457-3
ID636016      ID647706-1  ID638576-1   ID648715-3  ID419741-4  ID633723-4
ID632060-2    ID640854    ID630390-1   ID635680-1  ID639526    ID642613
ID641169      ID643752-1  ID637664-1   ID644041    ID630611-3  ID638495-1
ID624896-1    ID649571-2  ID636044-2   ID610122    ID352957-2  ID629085-2
ID640352-1    ID636700-1  ID627554-1   ID642449    ID572272    ID638838
ID367226-3    ID609200-1  ID637666-1   ID581746-6  ID637308-1  ID598289
ID610201-1    ID631688-8  ID608304-2   ID628337-2  ID646511-2  ID607246-8
ID641963      ID633391-2  ID641083-1   ID642090    ID550547-1  ID641013-6
ID639575-2    ID641360-1  ID643328     ID640829-1  ID651476-1  ID635116-3
ID625892-1    ID639406    ID632504-2   ID638219    ID572234-1  ID639486-1
ID642926-1    ID489499-2  ID628164-4   ID648786-1  ID651106-1  ID642562
ID637561-2    ID621976-5  ID638779     ID635126-2  ID618332-3  ID611691-6
ID635933      ID636613    ID508113-2   ID626594-3  ID640510-2  ID638556-1
ID642983-2    ID632668-6  ID637094     ID631715-2  ID626861-1
```

##### CONFIGURATION
- nothing of note

##### OS / ENVIRONMENT
N/A - REST issue is from BigIP

##### SUMMARY

REST returned a HTTP 400 Unexpected Error as described below.

Secondarily, I'd love to have a way to accomplish changes to member priority groups without using bigip_command, but that's another issue entirely.

##### STEPS TO REPRODUCE
Running playbook with repeated bigip_command tasks as below:
Have not been able to replicate beyond the initial instance.
Note that example task below has been edited for privacy.

```
  - name: Update pool member priority groups for poolA
    bigip_command:
      server: "{{ inventory_hostname }}"
      commands:
      - "modify ltm pool poolA members modify { {{ item.host }}:{{item.port }} { priority-group {{ item.pri }} } }"
    with_items:
    - { host: hostA.local.com, port: 12345, pri: 3}
    - { host: hostB.local.com, port: 12345, pri: 1}
```

##### EXPECTED RESULTS

On most invocations of the playbook, all of the bigip_command tasks return succcess. 

##### ACTUAL RESULTS

During one run (so far) a REST 400 has aborted the run. Note that the command output has been edited for privacy.

```
failed: [lv-bigip-100.local.com] (item={u'host': u'hostA.local.com', u'pri': 1, u'port': 12345}) => {"failed": true, "item": {"host": "hostA.local.com", "port": 12345, "pri": 1}, "msg": "400 Unexpected Error: Bad Request for uri: https://lv-bigip-100.local.com:443/mgmt/shared/authn/login\nText: u'{\"code\":400,\"message\":\"user 3d0fe353-f090-3dee-aa72-c27c3b2d36fd has reached maximum active login tokens\",\"originalRequestBody\":\"{\\\\\"username\\\\\":\\\\\"xxxxxxx\\\\\",\\\\\"loginProviderName\\\\\":\\\\\"tmos\\\\\",\\\\\"generation\\\\\":0,\\\\\"lastUpdateMicros\\\\\":0}\",\"referer\":\"10.2.3.4\",\"restOperationId\":10830311,\"kind\":\":resterrorresponse\"}'"}
```

**Comment by caphrim007** [comment]
ack'ed. Seems the f5-sdk and/or the ansible modules are not cleaning up tokens and when enough modules are run, you may exhaust them.

**Comment by loudsong** [comment]
Thought I'd mention that I ran into this today, and was able to reproduce it several times. I'll dig into it.

**Comment by caphrim007** [comment]
@earchibald @loudsong I'm told by the REST maintainers that this shouldnt happen with an "admin" level account. Can you say whether this is the case or not?

**Comment by loudsong** [comment]
@caphrim007 I was using the built-in 'admin' account on the BIG-IP.

**Comment by itblaked** [comment]
Hi, I encountered the same issue with v13 VE when using bigip_pool, bigip_node, bigip_virtual_server, bigip_pool_member. Using the default admin account.

**Comment by earchibald** [comment]
Sorry, I haven't been engaging with this issue for a while - I had enough trouble with our specialized requirements that I've paused my Ansible/F5 integration while I focus on other priorities.

I'm glad that @loudsong and @itblaked are able to provide some useful information.

**Comment by caphrim007** [comment]
@earchibald this is still a valid issue. we're aware of the problem but have not yet devised an elegant solution.

**Comment by itblaked** [comment]
Ok, thanks for the update @caphrim007. Is there an easy way to clear the tokens via rest or ssh/tmsh as a workaround?

**Comment by alonsocamaro** [comment]
Hi, 

I'm facing the same issue

A workaround would be greatly appreciated in the meantime this is resolved

Thanks

      Ulises

**Comment by shortenmark** [comment]
You can send an authenticated HTTP DELETE request to the resource below to delete all authentication tokens::

https://<host>/mgmt/shared/authz/tokens

There are more targeted URLS that you can use to specify a given UUID;

see https://devcentral.f5.com/wiki/iControl.Authentication_token_resource_API.ashx

**Comment by caphrim007** [comment]
@alonsocamaro @earchibald @loudsong @itblaked 

I wanted to talk about the technical problems in the case that it stirs someone's brain in response, because on the surface the solution sounds easy, but in practice it's less so. I'm looking for opinions on what is the most acceptable mediation of risk that we can put in place.

The first problem is that Ansible Tasks are intended to be separated from each other. Therefore its often a no-no to have anything in place that shares state across them. To a certain degree we can bend this rule because one could argue that fact-caching is doing this, to which i would respond "yes but sorta not". The fact caching is 

Storing the token on disk is not a good solution as it essentially leaves your password, in clear text, on the Ansible controller. This token is technically short lived (20 minutes), but it can be used to renew itself for up to 10 hours, so there is a period of time there where a valid token will be available to read in plain text.

The F5-SDK's token auth (`token=True`) is an illusion. When it is doing is re-using the token in the header for all subsequent requests after the first instead of each request. However the moment you create a new `ManagementRoot` object, it essentially creates a brand new token on the system. It never re-uses anything, and there's no way to tell it to re-use anything because it doesn't persist any of that to disk. Likewise, BIG-IP requires authentication of some means to even see the tokens that currently exist. So the SDK would need to

* store the token locally to be able to set it in a subsequent request after `ManagementRoot` is destroyed
or
* use password authentication on subsequent creations of the `ManagementRoot` object. Thus rendering "token auth" meaningless. Consider Ansible tasks that run in quick succession.

Some other things.

* Ansible Vault is not available to modules. Vault decrypts early in the process and at that time we already have "unencrypted" versions of the data exposed to the modules. Additionally, since modules are (by default) intended to run remotely, modules have no secondary access to Vault to read or write data.

* Fact caching is a potential solution for the storage layer of the problem, except it's not enabled by default, and for good reason. Cache maintenance is one of those impossible problems to get right, and what is "right" is specific to your local environment. I'm interested in a general purpose solution to this problem.

* Using token auth **still** requires that a user/password combo be supplied to every task. The reason is because in the event that an existing token expires midway through a task, you're probably going to want us to be able to renew that token and not outright fail your task. Imagine requiring `retries` and `delay`s on every task to cover the above scenario. Presumably that's a poor user experience

* One option is providing a `lookup` for a token. This would allow you to decrease the total number of tokens that are issued for each playbook. The problem you're seeing here is due to BIG-IP exhausting its capacity for new tokens. But these tokens are expiring all the time, so using something like a `lookup` to give you one token per `Play` would eliminate getting a token for every `Task`. It's not a fix though, only a bandage because the underlying issue is still there. Lookups could be used in the form of `environment` fields, `token: ....` parameters, or blocks using a `with_*` syntax.

* We could provide a `bigip_token` module. This would consume your standard user/pass/server/port parameters and return a token for storing in a `result` and for use later in subsequent tasks; think `token: ..` instead of `user`, `password`. The downside of this approach is that you need to add a new task at the beginning of each play that (IMHO) shouldn't be required. You've essentially exposed an implementation detail of the system. However, you could use it to also cleanup existing tokens I guess.

Leave comments and opinions in replies to this issue.

**Comment by shortenmark** [comment]
Whilst I personally quite like the idea of a bigip_token module; given the constraints you have outlined above; is it possible to just flush the assigned token upon successful task completion as a logical step within the bigip_command module. This approach to me seems the most modular, any caching or re-use of tokens adds additional complexity around checking for and acquiring new tokens if one does not already exist.

This would also negate the need to acquire a new token for every retry of the task. I'm not sure what overhead this would incur on the Big-IP if there are a lot of calls to that module but given we are getting a new token for each call at the moment I can't see this being an issue.

**Comment by caphrim007** [comment]
@shortenmark yes, cleaning up the existing token is one approach that can be implemented.

We still need to address the other concern that some people have expressed of repeatedly sending the user/pass over the wire if it is (technically) not necessary when using a long-lived token.

The solution to that problem is more complicated, because Ansible has no such way to express this at, for example, the play level or (even worse) the "plays happening across the next 10 hours even if there is a large delay between running them" level.

I think some people think that BIGIP has a token scheme like something you might get from Github or other webservice that is a life-time "API token". This is not the case. There is a maximum time limit to BIGIP tokens and, of course, people want us to maximize that valid lifetime.

It's technically possibly, but it's going to require more work that is separate for what is being asked for in this ticket.

**Comment by shortenmark** [comment]
I agree, this is why I like the idea of a bigip_token module, however I dread to think of the added complexity in each module needing an argument to tell it to use a token if one exists through another potential module, if not get a token, or just use the existing method for those users that do not need or want to use a long lived token...

At the moment it is possible to remediate the issue of excessive tokens being created using REST commands already available, just requires the logic being written into the playbook, bit of a hack but works in the absence of something neater.

**Comment by alonsocamaro** [comment]
Adding to the comment above, the curl command for the first option [mentioned above](https://github.com/F5Networks/f5-ansible/issues/269#issuecomment-329181271) would be:

curl -k https://<bigip>/mgmt/shared/authz/tokens -u admin:<admin password> -X DELETE

**Comment by alonsocamaro** [comment]
Having a chat with John Allen we wonder if it wouldn't be better to handle auth at the f5-sdk library so it would be transparent to the f5-sdk library consumers.

Also, is there a way of using basic auth? I haven't found a way of instructing this. Maybe I haven't dig enough.

**Comment by caphrim007** [comment]
It would be more difficult to add it to the sdk as that's not necessary the purpose of it.

No, there is no way of using basic auth "for every request". The SDK takes a 'token' argument to its ManagementRoot that will cause the Token header to be generated. All the subsequent requests using the management object will use the token.

We do not allow control of that via Ansible because **ALL** BIG-IPs will be turning off basic auth in the future and requiring token auth.

It is likely that we can have Ansible so this for us, but it is not yet something I have worked on. The way one would do it is to embed this logic in the f5_utils.py module_utils file. Other required steps need to be included to manage temporary files between invocations of Ansible.

Ansible has a upcoming feature (though I forget in which release) that will allow for some sort of persistence of non-ssh "stuff", but it has not landed yet, been documented, tested, ensured that it's stable yada yada. In the meantime, we need to do it on our own.

**Comment by caphrim007** [comment]
Token re-use across sessions is being progressively added.

**Comment by itblaked** [comment]
👍

**Comment by chen23** [comment]
FYI, botocore does do some goofy things to cache/refresh credentials: https://github.com/boto/botocore/blob/cb8d4b87fbe13b42653017b74c9a5f4097ae740c/botocore/credentials.py#L194

Something like this could be in f5-ansible or down in f5-icontrol-rest-python.  Related to https://github.com/F5Networks/f5-ansible/issues/322 

Things that could be nice:

A. admin/password transparently changed to a token and refreshed for end-user (heavy lifting done by code)
B. expose a token auth as an alternative to username/password (end-user expected to generate a token first, easy to code, harder for end-user to generate token, token can still expire unless refreshed by end-user, provides migration path to A?)

A would be preferable (but could take longer/break existing code/playbooks), but B could provide an interim solution where you create a wrapper script to dynamically fetch a token before executing each playbook and/or environment variable... 

My $0.02.

**Comment by webminster** [comment]
This is definitely a +1 for us as well, our automation fails randomly due to this.   Would like to see this a priority.

**Comment by jbidinger** [comment]
A big problem for us as well.

**Comment by amsinha2** [comment]
+1  we hit this a lot, and it appears that the maximum tokens is shared across a pair

**Comment by delgadillo22** [comment]
Low if this is this a corner case situation with a workaround

**Comment by caphrim007** [comment]
@amsinha2 @jbidinger @webminster @chen23 @itblaked @alonsocamaro @shortenmark @earchibald @loudsong 

A fix for this should now be in 2.5 for all f5 modules. All new modules include the fix too

**Comment by jbidinger** [comment]
Thank you. We appreciate it very much.

**Comment by alonsocamaro** [comment]
@caphrim007

**Comment by ksbarnt** [comment]
I'm running into this issue when running Ansible 2.7.5 against a BIG-IP VE box running 13.1.1.3 (also previously 13.1.1) with a lab license. Is it possibly the lab license that's causing me to still have issues? I'm using a user with full admin privileges.

**Comment by pauljolsen** [comment]
I get this error after 100 plays with 2.7.  In 2.8 the code that deletes the tokens as it goes appears to show up here https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network/f5/icontrol.py#L243

**Comment by andrenrwn** [comment]
There is a limit to the number of concurrent tokens issued for each user for REST referenced here:

[K44237337: iControl REST: Limitation on a number of authentication tokens per user](https://my.f5.com/manage/s/article/K44237337)

[K59142111: Error Message ''user 21232f29-7a57-35a7-8389-4a0e4a801fc3 has reached maximum active login tokens'' Is Returned Unexpectedly](https://my.f5.com/manage/s/article/K59142111)

A possible solution would be ensuring that tokens are removed once the connection is done

[K02126317: How to delete/revoke a user REST API token](https://my.f5.com/manage/s/article/K02126317)

There are ways to look at what tokens are issued for a specific BIG-IP user:
[K13304470: Using a shell function to retrieve authentication tokens from BIG-IPs and BIG-IQs](https://my.f5.com/manage/s/article/K13304470)
